components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,328 @@
+        return Err("Serialization must be even".into());
+    }
+    serialized_vec.shrink_to_fit(); // necessary if the length < capacity
+    Ok(serialized_vec)
+}
+
+/// UnicodeSet membership wrapper
+///
+/// Provides exposure to membership functions and constructors from serialized UnicodeSets
+/// and predefined ranges.
+//#[derive(Copy, Clone, Debug, Eq)]
+pub struct UnicodeSet {
+    // If we wanted to use an array to keep the memory on the stack, there is an unsafe nightly feature
+    // https://doc.rust-lang.org/nightly/core/array/trait.FixedSizeArray.html
+    // Allows for traits of fixed size arrays
+    inv_list: Vec<u32>,
+}
+
+impl UnicodeSet {
+    /// Returns Result of UnicodeSet from serialized string
+    ///
+    /// Returns an error if the serialized string fails to parse.
+    /// The serialized string requires starting capacity integer, followed by space delimited
+    /// integer code points. There must be an even number of elements (not including the
+    /// capacity int), and must be in ascending sorted order.
+    ///
+    /// Example String: `"4 0 5 10 15"` designates a capacity of size `4`, followed by 2 ranges
+    /// The ranges are `{0, 4}` and `{10, 14}` inclusive
+    pub fn new(serialize: &str) -> Result<UnicodeSet, Box<dyn Error>> {
+        match parse_serial_string(serialize) {
+            Ok(serialize) => Ok(UnicodeSet {
+                inv_list: serialize,
+            }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Returns Result of UnicodeSet from a single pair of integers defining a range
+    ///
+    /// `start`: inclusive, `end`: exclusive
+    ///
+    /// Returns an error if the range is invalid (out of order and out of bounds).
+    ///
+    /// Example Call: `UnicodeSet::from_range(&0, &15)`
+    pub fn from_range(start: &u32, end: &u32) -> Result<UnicodeSet, Box<dyn Error>> {
+        if start > end {
+            return Err("Range is out of order".into());
+        }
+        if start < &CODEPOINT_MIN || end > &CODEPOINT_MAX {
+            return Err("Range is out of bounds".into());
+        }
+        Ok(UnicodeSet {
+            inv_list: vec![*start, *end],
+        })
+    }
+
+    /// Returns UnicodeSet spanning entire Unicode range
+    ///
+    /// The range spans from `0x0 -> 0x10FFFF` inclusive
+    pub fn all() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, CODEPOINT_MAX + 1],
+        }
+    }
+
+    /// Returns UnicodeSet spanning BMP range
+    ///
+    /// The range spans from `0x0 -> 0xFFFF` inclusive
+    pub fn bmp() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, BMP_MAX + 1],
+        }
+    }
+
+    /// Returns an `Iter` of start and stop `u32` points of the UnicodeSet
+    pub fn iter(&self) -> Iter<u32> {
+        self.inv_list.iter()
+    }
+
+    /// Returns the cardinality of the UnicodeSet
+    ///
+    ///
+    pub fn size(&self) -> Result<usize, Box<dyn Error>> {
```

Comment:
It seems unconventional that a simple query such as `size()` can error out.  Is there a way to push a check for this error condition to some earlier point in the set's lifetime so that you don't need to error out here?

For example, if your constructor or mutator methods check for `set.inv_list.len() >=2`, then you can turn this into an `assert!()` and remove the need to return `Result` here.

In general try to make creating objects with invalid internal state impossible.  If it is inevitable that `inv_list` is not initialized correctly at some point in *its* lifetime, perhaps you need a builder which would have a `build()` method that includes well-formedness checks.

Reply:
I agree regarding the unconventional-ness of the error. Assuming the user does not initialize their own `UnicodeSet`, rather than using the constructor methods, there should be no error.

I just worry that a user may call: `UnicodeSet {inv_list: vec![10]}` rather than `UnicodeSet::new(...)`. Should the `UnicodeSet{inv_list: vec![]}` possibility not be exposed to the public?

Reply:
I realized this isn't possible as `inv_list` is already private. Ignore the above comment. Thanks!

Reply:
> I just worry that a user may call: `UnicodeSet {inv_list: vec![10]}` rather than `UnicodeSet::new(...)`. Should the `UnicodeSet{inv_list: vec![]}` possibility not be exposed to the public?

IMHO users must not be allowed to initialize your struct in a way that makes it internal state invalid.  If there are invalid values of `inv_list`, then inv_list must not be public to ensure that the type's invariants hold.

Reply:
agreed and changed.